addons/stock/models/stock_move_line.py

```diff
@@ -98,8 +98,52 @@ def onchange_product_id(self):
 
     @api.onchange('lot_name', 'lot_id')
     def onchange_serial_number(self):
+        res = {}
         if self.product_id.tracking == 'serial':
             self.qty_done = 1
+            move_lines_to_check = self._get_move_lines() - self._origin
+            message = move_lines_to_check._check_lot_exists()
+            if message:
+                res['warning'] = {'title': _('Warning'), 'message': message}
+        return res
+
+    @api.constrains('lot_id', 'lot_name', 'qty_done')
+    def _check_unique_serial_number(self):
+        move_lines_serial = self.filtered(lambda ml: ml.move_id.product_id.tracking == 'serial' and (ml.lot_id or ml.lot_name))
+        for ml in move_lines_serial:
+            move_lines_to_check = ml._get_move_lines() - ml
+            message = move_lines_to_check._check_lot_exists()
+            if message:
+                raise ValidationError(message)
+            if float_compare(ml.qty_done, 1.0, precision_rounding=ml.move_id.product_id.uom_id.rounding) == 1:
+                raise UserError(_(
+                    'You can only process 1.0 %s for products with unique serial number.') % ml.product_id.uom_id.name)
+            if ml.lot_name:
+                already_exist = self.env['stock.production.lot'].search(
+                    [('name', '=', ml.lot_name), ('product_id', '=', ml.product_id.id)])
+                if already_exist or float_compare(ml.qty_done, 1.0, precision_rounding=ml.move_id.product_id.uom_id.rounding) == 1:
+                    return _('You have already assigned this serial number to this product. Please correct the serial numbers encoded.')
+
+    def _get_move_lines(self):
```

Comment:
there's already a method named `_get_move_lines`on stock.move that return the move lines to consider for _quantity_done_compute, i suggest to change this method name for clarity sake